### DISCUSSION TITLE
Interesting Comparison Cases

### DISCUSSION THREAD

**Discussion by mastratton3** [moj-analytical-services/splink#1243] [General]
I've been working on a lot of medium sized record linkage tasks lately, generally in the range of 5-20k records joining to 5-20k records but some are larger. These are tasks like linking Hospital Certification Numbers to their NPIs or Ambulatory Surgical Center Certifications to their NPIs.

One comparison idea that continually comes up is the idea of a field that has very high predictive power when it's generally unique* but very low when shared with other entities. A couple examples of this:
1. Mailing Address - Generally high predictive power except in the case the address is a virtual mailbox.
2. Phone Number - Generally high predictive power except in the case of a centralized answering service (ie: 5 ASCs in the same region with common owners may have a single phone number)
3. Address & Suite Uniqueness - Similar to the above, when you have an address or suite that only has 1 entry in each dataset it's highly predictive but if you have the address/suite shared (Particularly if it's shared in the certification dataset) it significantly reduces the probability they're the same. 

*Generally Unique - It's hard to enforce a true "unique" constraint as you may have some noise in the data such as an ASC that closed and reopened with the same name, phone, etc... and thus has 2 certification numbers.... 

My current thought on handling this is:
Define a comparison level such as "One address is Likely Virtual Mailbox" in my address compare (Although even these aren't super simple to determine) prior to checking if the addresses match. Or checking if they match and then checking if the matched address is a virtual mailbox.

I realize this is a bit of a general question, just curious if anyone else has encountered anything similar.

**Comment by RobinL** [comment]
Not familiar with medical data, but this may have some similarity to some challenges we've encountered.

Here are some thoughts:

- If the true cardinality of a field is greater than the number of records, Splink will usually underestimate the importance of a field.  To give an example, if you were linking 100 records, and one of the fields was social security number, it would estimate that the probability that two _different_ people have the same social security number is one in a hundred.  (if you used [estimate_u_using_random_sampling](https://moj-analytical-services.github.io/splink/linker.html?h=estimate_u#splink.linker.Linker.estimate_u_using_random_sampling)).    This is because it uses the cardinality of the data it can see, and has no knowledge of the true underlying cardinality
- For the same reason, term frequency adjustments will underestimate the true value of a very high cardinality field on small datasets.
- One solution to this problem may be to:
  - Split a field that's a mix of 'truly unique' identifiers and 'not-so-unique' identifiers into two fields:  'mailing_address_common' and 'mailing_address_uncommon', where 'mailing_address_uncommon' contains the 'unique' mailing addresses and is null otherwise, and 'mailing_address_common' contains the 'non-unique' mailing addresses and is null otherwise
  - Manually set the u probability on the exact match level of mailing_address_uncommon to something very small that reflects the true cardinality of this field.  i.e. rather than 1/5_000, you might set it to 1/1_000_000
  - If your common virtual mailbox addresses have cardinality greater than the size of your dataset (say 1/20_000), you may wish to also manually set the u probability of the exact match level of  mailing_address_common

Another way of surmounting this problem is to calculate a term frequency table using the full dataset of mailing addresse - e.g. you may be linking 20k records to 20k, but presumably these are subsets of some larger dataset.  If you have access to the 'full dataset' of these addresses (which may be millions of records), you can compute a term frequency table on the full dataset, and register this.  It will then use the correct cardinality in the calculations
https://moj-analytical-services.github.io/splink/linker.html?h=compute_tf#splink.linker.Linker.compute_tf_table
https://github.com/moj-analytical-services/splink/blob/594cfa813e38592a43c15d05c5c3a4d96fac8c2a/splink/linker.py#L2912

**Comment by mastratton3** [comment]
Very helpful feedback - fwiw I'd say this is more of "business entity" data than healthcare data, they just happen to be healthcare businesses.

I really like your SSN example for demonstrating potential issues w/ calculating u.
